File: main/src/main/scala/org/clulab/dynet/FinalLayer.scala
package org.clulab.dynet

import edu.cmu.dynet.{ComputationGraph, Expression, ExpressionVector}

trait FinalLayer extends Saveable {
  def forward(inputExpressions: ExpressionVector,
              modifierHeadPairs: Option[IndexedSeq[ModifierHeadPair]],
              doDropout: Boolean)(implicit cg: ComputationGraph): ExpressionVector

  def inDim: Int
  def outDim: Int

  def loss(emissionScoresAsExpression: ExpressionVector, goldLabels: IndexedSeq[Label])(implicit cg: ComputationGraph): Expression

  def inference(emissionScores: Array[Array[Float]])(implicit cg: ComputationGraph): IndexedSeq[String]

  def inferenceWithScores(emissionScores: Array[Array[Float]])(implicit cg: ComputationGraph): IndexedSeq[IndexedSeq[(String, Float)]]
}

File: main/src/main/python/pytorch/finalLayer.py
import torch
import torch.nn as nn

class FinalLayer(nn.Module):

    def __init__(self):
        super().__init__()
        self.inDim = None
        self.outDim = None

    def forward(self, inputExpressions, headPositionsOpt, doDropout):
        raise NotImplementedError

    def loss(self, emissionScoresAsExpression, goldLabels):
        raise NotImplementedError

    def inference(self, emissionScores):
        raise NotImplementedError

    def inferenceWithScores(self, emissionScores):
        raise NotImplementedError